SSA/Projects/InstCombine/scripts/mlir-tool.py

```diff
@@ -0,0 +1,183 @@
+#!/usr/bin/env python3
+from xdsl.dialects.builtin import ModuleOp
+from xdsl.dialects.llvm import LLVM
+from xdsl.utils.exceptions import ParseError
+from xdsl.context import MLContext
+from xdsl.dialects import get_all_dialects
+from xdsl.dialects.llvm import FuncOp
+from xdsl.parser import Parser
+from xdsl.printer import Printer
+from xdsl.dialects.builtin import (
+    Builtin,
+    IndexType,
+    IntegerAttr,
+    IntegerType,
+    ModuleOp,
+    StringAttr,
+    i32,
+    i64,
+)
+import os
+import io
+from pathlib import Path
+from xdsl.printer import Printer
+
+# Initialize the MLIR context and register the LLVM dialect
+ctx = MLContext(allow_unregistered=True)
+ctx.load_dialect(LLVM)
+ctx.load_dialect(Builtin)
+
+directory = os.fsencode("../vcombined-mlir")
+
+
+allowed_names = {
+    "llvm.return",
+    "llvm.mul",
+    "llvm.add",
+    "llvm.sub",
+    "llvm.shl",
+    "llvm.and",
+    "llvm.or",
+    "llvm.xor",
+    "llvm.mlir.constant",
+    "llvm.lshr",
+    "llvm.ashr",
+    "llvm.urem",
+    "llvm.srem",
+    "llvm.add",
+    "llvm.mul",
+    "llvm.sub",
+    "llvm.sdiv",
+}
+allowed_unregistered = set()  # {
+
+
+# "llvm.icmp"
+# }
+def allowed(op):
+    return (
+        hasattr(op, "sym_name")
+        or (
+            op.name == "builtin.unregistered"
+            and op.op_name.data in allowed_unregistered
+        )
+        or (op.name in allowed_names)
+    )
+
+
+def show(block):
+    output = io.StringIO()
+    p = Printer(stream=output)
+    block.print(p)
+    contents = output.getvalue()
+    output.close()
+    return contents
+
+
+def showr(region):
+    output = io.StringIO()
+    p = Printer(stream=output)
+    p.print_region(region)
+    contents = output.getvalue()
+    output.close()
+    return contents
+
+
+def size(func):
+    return sum(1 for _ in func.walk())
+
+
+def read_file(file_name):
+    with open(file_name, "r") as f:
+        return f.read()
+
+
+for file in os.listdir(directory):
+    filename = os.fsdecode(file)
+    print(filename)
+    stem = "g" + filename.split(".")[0].replace("-", "h")
+    output = ""
+    data1 = read_file(os.path.join("../vcombined-mlir"))
+    data2 = read_file(os.path.join("../vbefore-mlir"))
+    parser = Parser(ctx, data1)
+    try:
+        module = parser.parse_module()
+        parser2 = Parser(ctx, data2)
+        module2 = parser2.parse_module()
```

Comment:
```suggestion
    with open(os.path.join("../vcombined-mlir"))) as f1:
        module1 = Parser(ctx, f1).parse_module()
    with open(os.path.join("../vbefore-mlir"))) as f2:
        module2 = Parser(ctx, f2).parse_module()
    try:
```

Reply:
It seems that xDSL Parser needs a string, not a file

https://github.com/xdslproject/xdsl/blob/9b43500e6526196b31b5a27a413af9475c1b71e0/xdsl/parser/core.py#L91C5-L102C1

Reply:
Then can you read in the with block, please?

Reply:
> Then can you read in the with block, please?
What do you mean by this?